pkg/roll/run_update.go

```diff
@@ -0,0 +1,305 @@
+	}
+	if newNodes.healthy+oldNodes.healthy <= u.MinimumReplicas {
+		return fmt.Errorf("minimum (%d) is not satisfied by existing healthy nodes (%d old, %d new), update would block immediately", u.MinimumReplicas, oldNodes.healthy, newNodes.healthy)
+	}
+	return nil
+}
+
+func (u update) Run(quit <-chan struct{}) error {
+	newFields, err := u.rcs.Get(u.NewRC)
+	if err != nil {
+		return err
+	}
+	hChecks := make(chan map[string]health.Result)
+	hErrs := make(chan error)
+	hQuit := make(chan struct{})
+	defer close(hQuit)
+	go u.hcheck.WatchService(newFields.Manifest.ID(), hChecks, hErrs, hQuit)
+
+ROLL_LOOP:
+	for {
+		select {
+		case <-quit:
+			return nil
+		case <-hErrs:
+			return err
+		case checks := <-hChecks:
+			newNodes, err := u.countHealthy(u.NewRC, checks)
+			if err != nil {
+				return err
+			}
+			oldNodes, err := u.countHealthy(u.OldRC, checks)
+			if err != nil {
+				return err
+			}
+
+			if newNodes.desired > newNodes.healthy {
+				// we assume replication controllers do in fact "work", ie healthy
+				// and current converge towards desired
+				// in this case, they have not yet met, so we block
+				u.logger.WithFields(logrus.Fields{
+					"old": oldNodes,
+					"new": newNodes,
+				}).Debugln("Blocking for more healthy new nodes")
+				break
+			}
+			if newNodes.desired >= u.DesiredReplicas {
+				// note that we only exit the loop AFTER the desired nodes
+				// become healthy - this ensures that, when the old RC is
+				// cleaned up, we do not accidentally remove the nodes we just
+				// assigned to the new RC
+				// normal execution of this algorithm will never cause
+				// newDesired > u.Desired, but if Run is resuming from an unusual
+				// situation, we don't want to get stuck
+				u.logger.WithFields(logrus.Fields{
+					"old": oldNodes,
+					"new": newNodes,
+				}).Debugln("Upgrade complete")
+				break ROLL_LOOP
+			}
+			next := algorithm(oldNodes.desired, newNodes.desired, u.DesiredReplicas, u.MinimumReplicas)
+			u.logger.WithFields(logrus.Fields{
+				"old":  oldNodes,
+				"new":  newNodes,
+				"next": next,
+			}).Infoln("Undergoing next update")
+
+			if next > 0 {
+				if u.DeletePods {
+					u.rcs.AddDesiredReplicas(u.OldRC, -next)
+				}
+				u.rcs.AddDesiredReplicas(u.NewRC, next)
+			}
+		}
+	}
+
+	// rollout complete, clean up old RC
+	u.logger.NoFields().Infoln("Cleaning up old RC")
+	if err := u.rcs.SetDesiredReplicas(u.OldRC, 0); err != nil {
+		return err
+	}
+	if err := u.rcs.Enable(u.OldRC); err != nil {
+		return err
+	}
+	if err := u.rcs.Delete(u.OldRC, false); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (u update) lock(session string) error {
+	if success, err := u.rcs.LockWrite(u.NewRC, session); err != nil {
+		return err
+	} else if !success {
+		return fmt.Errorf("could not lock new %s", u.NewRC)
+	}
+
+	if success, err := u.rcs.LockWrite(u.OldRC, session); err != nil {
+		return err
```

Comment:
Well if session renewal fails, the ConsulSessionManager will immediately send a `""` to indicate that it can no longer guarantee the session's liveness. What i"m most interested in is if a program holding a session gracefully exits - it should destroy its session to free the locks it was holding immediately rather than waiting for up to 15s TTL.

Although in practice that doesn't matter, because nobody would try to claim the lock until their watches expired.